Google.PowerShell/Common/ReportCmdletResults.cs

```diff
@@ -0,0 +1,189 @@
+        }
+
+        public void ReportFailure(string cmdletName, string parameterSet, int errorCode)
+        {
+            Report(EventRecord.Create(cmdletName, parameterSet, errorCode));
+        }
+
+        protected void Report(EventRecord record)
+        {
+            // Overwrite older events as if it were a circular buffer.
+            _events[_eventsRecorded % kMaxEvents] = record;
+            _eventsRecorded++;
+        }
+
+        /// <summary>
+        /// Clears all history of events.
+        /// </summary>
+        public void Reset()
+        {
+            _eventsRecorded = 0;
+            _events = new EventRecord[kMaxEvents];
+        }
+
+        /// <summary>
+        /// Returns whether or not an event with the given makeup has been recorded.
+        /// </summary>
+        public bool ContainsEvent(string cmdletName, string parameterSet, int errorCode = Int32.MinValue)
+        {
+            var expectedRecord = EventRecord.Create(cmdletName, parameterSet, errorCode);
+            for (int i = 0; i < _events.Length; i++)
+            {
+                if (_events[i].Equals(expectedRecord))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+    }
+
+    /// <summary>
+    /// Type of analytic event.
+    /// </summary>
+    public enum AnalyticsEventCategory
+    {
+        CmdletInvocation,
+        Provider
+    }
+
+    /// <summary>
+    /// Reports PowerShell cmdlet results to Google Analytics.
+    /// </summary>
+    public class GoogleAnalyticsCmdletReporter : IReportCmdletResults
+    {
+        // Google Analytics Property ID, see di-metrics.
+        private const string PropertyId = "UA-36037335-1";
+        // Application name to report to GA.
+        private const string ApplicationName = "google-cloud-powershell";
+        // Version of the Google.PowerShell assembly.
+        private static readonly Lazy<string> s_appVersion = new Lazy<string>(() => Assembly.GetExecutingAssembly().GetName().Version.ToString());
+
+        /// <summary>
+        /// Analytics reporter object.
+        /// </summary>
+        private Lazy<AnalyticsReporter> _reporter = null;
+
+        /// <summary>
+        /// Category for all analytic events reported through this instance.
+        /// </summary>
+        private AnalyticsEventCategory _category = AnalyticsEventCategory.CmdletInvocation;
+
+        public GoogleAnalyticsCmdletReporter(string clientID, AnalyticsEventCategory analyticsCategory)
+        {
+            // We report results to Google Analytics if the user has opted-into Cloud SDK metric reporting. However, we
+            // don't want to report data during development. (e.g. when running unit tests.)
+            string envVar = Environment.GetEnvironmentVariable("DISABLE_POWERSHELL_ANALYTICS");
+            bool debugFlag = (envVar == "TRUE");
```

Comment:
Suggestion, maybe less stringent check is better. Just check that is not empty. Checking for only TRUE can make it a bit error prone.